src/coord/rmr/chan.c

```diff
@@ -109,6 +112,89 @@ void *MRChannel_Pop(MRChannel *chan) {
   return ret;
 }
 
+void *MRChannel_PopWithTimeout(MRChannel *chan, const struct timespec *abstimeMono, bool *timedOut) {
+  *timedOut = false;
+
+  // If no timeout specified, behave like regular Pop
+  if (!abstimeMono) {
+    return MRChannel_Pop(chan);
+  }
+
+#if defined(__APPLE__) && defined(__MACH__)
+  // macOS: use pthread_cond_timedwait_relative_np with relative timeout
+  // Must recalculate remaining time on each iteration to handle spurious wakeups
+  pthread_mutex_lock(&chan->lock);
+  while (!chan->size) {
+    if (!chan->wait) {
+      chan->wait = true;  // reset the flag
+      pthread_mutex_unlock(&chan->lock);
+      return NULL;
+    }
+
+    // Recalculate remaining time on each iteration to handle spurious wakeups
+    struct timespec nowMono, remaining;
+    clock_gettime(CLOCK_MONOTONIC, &nowMono);
+    rs_timerdelta((struct timespec *)abstimeMono, &nowMono, &remaining);
+
+    // If already past the deadline, timeout immediately
+    if (remaining.tv_sec == 0 && remaining.tv_nsec == 0) {
+      *timedOut = true;
+      pthread_mutex_unlock(&chan->lock);
+      return NULL;
+    }
+
+    int rc = pthread_cond_timedwait_relative_np(&chan->cond, &chan->lock, &remaining);
```

Comment:
We should implement a generic `pthread_cond_timedwait` and handle the different OSs there, and keep the channel popping logic simple. I think this is currently hard to maintain and follow